Add RSSI and HEV sensors to LIFX devices

Djelibeybi opened the pull request:
## Proposed change

Adds RSSI signal strength sensors to all LIFX devices and HEV sensors to LIFX Clean devices. 

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [X] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Djelibeybi committed:
Add RSSI and HEV sensors to LIFX devices

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Add RSSI and HEV sensors to LIFX devices

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
@bdraco this is my attempt at adding sensor entites. I picked one that is available to all bulbs and some that are only available to a subset of bulbs. In draft fo review so I can fix anything I got wrong. :)

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        """Return the native unit of measurement for this sensor."""
+        return cast(str, self.entity_description.native_unit_of_measurement)
+
+    @property
+    def native_value(self) -> str:
+        """Return the value from the bulb."""
+        return cast(str, self._attr_native_value)
+
```
```suggestion
```

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+    def native_unit_of_measurement(self) -> str:
+        """Get the RSSI unit of measurement."""
+        return self.coordinator.get_rssi_unit_of_measurement()
+
+    @property
+    def native_value(self) -> int:
+        """Get the current RSSI value."""
+        return cast(int, self._attr_native_value)
```
```suggestion
```

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        """Initialise a LIFX HEV sensor."""
+        super().__init__(coordinator=coordinator, description=description)
+
+    @property
+    def native_unit_of_measurement(self) -> str:
+        """Return the native unit of measurement for this sensor."""
+        return cast(str, self.entity_description.native_unit_of_measurement)
+
```
```suggestion
```

Set `self.entity_description = description` instead

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+    state_class=SensorStateClass.MEASUREMENT,
+    device_class=SensorDeviceClass.SIGNAL_STRENGTH,
+    has_entity_name=True,
+    icon="mdi:wifi",
```
This should be disabled by default so it doesn't generate too many updates in the state machine
```suggestion
    icon="mdi:wifi",
    entity_registry_enabled_default=False
```

Djelibeybi replied:
Heh, it was initially, but I turned that off for testing purposes so I didn't have to manually turn it on for 60 bulbs. 😂

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the LIFX device info."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.coordinator.serial_number)},
+            connections={(dr.CONNECTION_NETWORK_MAC, self.coordinator.mac_address)},
+            manufacturer="LIFX",
+            name=self.name,
+        )
```
You can set this as `self._attr_device_info = `

bdraco commented on homeassistant/components/lifx/coordinator.py:
```diff
@@ -156,3 +202,43 @@ async def async_set_color_zones(
             )
         )
+
+    async def async_update_wifi_info(self) -> None:
+        """Get the bulb's wifi signal strength."""
+        _LOGGER.debug("Updating wifi signal strength for %s", self.device.label)
+        try:
+            wifi_info = await async_execute_lifx(self.device.get_wifiinfo)
+            self._rssi = int(math.floor(10 * math.log10(wifi_info.signal) + 0.5))
+        except asyncio.TimeoutError as ex:
+            raise UpdateFailed(
+                f"Failed to fetch RSSI sensor data from device: {self.device.ip_addr}"
+            ) from ex
+
```
We shouldn't poll this unless at least one entity is enabled that needs it

Djelibeybi replied:
Yeah, I wasn't sure how to work that out from the coordinator. All my attempts to override `async_update` failed.

bdraco replied:
You could keep track of it by adding having the entity tell the coordinator it wants it in `async_added_to_hass` and tell it, it no longer wants it in `async_will_remove_from_hass`

Similar example https://github.com/home-assistant/core/blob/6da25c733e014624564140a761af5201e3bf0adc/homeassistant/components/unifiprotect/entity.py#L217

bdraco commented on homeassistant/components/lifx/coordinator.py:
```diff
@@ -156,3 +202,43 @@ async def async_set_color_zones(
+                f"Failed to fetch RSSI sensor data from device: {self.device.ip_addr}"
+            ) from ex
+
+    async def async_update_hev_cycle(self) -> None:
```
We shouldn't poll this unless at least one entity is enabled that needs it

Djelibeybi replied:
This is only available to HEV bulbs and they would all have it enabled by default, so if this is called, it's needed.

bdraco replied:
Users can disable the entities

bdraco commented on homeassistant/components/lifx/coordinator.py:
```diff
@@ -156,3 +202,43 @@ async def async_set_color_zones(
+            hev_cycle: StateHevCycle = await async_execute_lifx(
+                self.device.get_hev_cycle
+            )
+            self._hev_cycle = {
+                HEV_CYCLE_DURATION: hev_cycle.duration,
+                HEV_CYCLE_REMAINING: hev_cycle.remaining,
+                HEV_CYCLE_LAST_POWER: hev_cycle.last_power,
+            }
+
+            last_result: StateLastHevCycleResult = await async_execute_lifx(
+                self.device.get_last_hev_cycle_result
+            )
+            self._hev_cycle[HEV_CYCLE_LAST_RESULT] = (
+                str(last_result.result_str)
+                .title()
+                .replace("_", " ")
+                .replace("Homekit", "HomeKit")
+                .replace("Lan", "LAN")
+            )
+            _LOGGER.debug(self._hev_cycle)
```
Please move code out of the coordinator that won't raise.
`
            hev_cycle: StateHevCycle = await async_execute_lifx(
                self.device.get_hev_cycle
            )
            last_result: StateLastHevCycleResult = await async_execute_lifx(
                self.device.get_last_hev_cycle_result
            )
`

Then process it in an `else`

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        key=HEV_CYCLE_LAST_POWER,
+        name="Clean Cycle Last Power",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        state_class=None,
+        device_class=None,
```
```suggestion
```

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        key=HEV_CYCLE_LAST_RESULT,
+        name="Clean Cycle Last Result",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        state_class=None,
+        device_class=None,
```
```suggestion
```

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        state_class=None,
+        device_class=None,
+        native_unit_of_measurement=None,
```
```suggestion
```

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DURATION,
+        native_unit_of_measurement=TIME_MINUTES,
+        icon="mdi:timer",
```
```suggestion
```

We normally don't set an `icon` if the `device_class` is set. https://developers.home-assistant.io/docs/core/entity#advanced-properties

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        state_class=None,
+        device_class=SensorDeviceClass.DURATION,
+        native_unit_of_measurement=TIME_MINUTES,
+        icon="mdi:timer",
```
```suggestion
```


We normally don't set an `icon` if the `device_class` is set. https://developers.home-assistant.io/docs/core/entity#advanced-properties

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        key=HEV_CYCLE_DURATION,
+        name="Clean Cycle Duration",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        state_class=None,
```
```suggestion
```

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,208 @@
+    """Set up LIFX from a config entry."""
+    domain_data = hass.data[DOMAIN]
+    coordinator: LIFXUpdateCoordinator = domain_data[entry.entry_id]
+    sensors: list[LifxRssiSensorEntity | LifxHevSensorEntity] = [
```
```suggestion
    sensors: list[LifxSensorEntity] = [
```

bdraco commented on homeassistant/components/lifx/util.py:
```diff
@@ -25,11 +26,14 @@
 FIX_MAC_FW = AwesomeVersion("3.70")
 
 
+_LOGGER = logging.getLogger(__name__)
```
stale debug ?

Djelibeybi replied:
No. Moving `_LOGGER` to `.const` makes all the logs look like they come from `homeassistant.components.lifx.const` and not the actual file in which it is called. It makes more sense to define it for each file to make debugging a little bit easier.

bdraco commented on homeassistant/components/lifx/coordinator.py:
```diff
@@ -156,3 +202,43 @@ async def async_set_color_zones(
+                self.device.get_last_hev_cycle_result
+            )
+            self._hev_cycle[HEV_CYCLE_LAST_RESULT] = (
+                str(last_result.result_str)
+                .title()
+                .replace("_", " ")
+                .replace("Homekit", "HomeKit")
+                .replace("Lan", "LAN")
```
It would be nice if this was a separate function

bdraco commented:
Will try to give you some more feedback tomorrow.  I'm way behind as I've had two A/C units fail this week @ home which is not good in a Texas summer !

Djelibeybi committed:
WIP: add diagnostic sensors to LIFX bulbs

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
WIP: add diagnostic sensors to LIFX bulbs

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
I have it working as requested, but I'm not sure if I did it properly or not. Decided to stop here and give you a chance to review again before doing anything more.

bdraco commented on homeassistant/components/lifx/coordinator.py:
```diff
@@ -75,9 +102,79 @@ def mac_address(self) -> str:
+                    f"Failed to fetch HEV cycle status from device: {self.device.label}() {self.device.ip_addr})"
+                ) from ex
+
+        @Throttle(timedelta(seconds=60))
```
We try to avoid using Throttle in new code

If we need different intervals, usually we use multiple coordinators or write a custom coordinator. 

It's a bit more difficult here since we need a shared lock but the advantage is if you have multiple coordinators you don't have to worry about subscription management since the coordinators don't poll if there are no subscribers

Djelibeybi replied:
Strictly speaking, we don't need different intervals but then we also don't need to poll for sensor data every 10 seconds either. I'm more worried about the amount of data potentially being stored by the logger for no good reason if we poll too often.

Djelibeybi replied:
We probably don't need to share a lock either: that's only really needed for discovery. Once the transport is unicast UDP, the bulbs tend to respond pretty quickly and can handle up to 20 pkts/second. 

As a reference, the Tile animations generated by Photons send 320 HSBK values every 0.075 seconds. :)

Djelibeybi replied:
We would be better suited with a global limit of 2-4 inflight (unacknowledged) packets per endpoint, which is automatic in Photons but would have to be added to aiolifx.

bdraco replied:
Reducing the polling frequency makes a lot of sense here. If you have a lot of bulbs the additional wifi traffic will make a difference on how much reliable your 2.4g network is.

Djelibeybi replied:
Agreed. And it would reduce the amount of potential state data being stored by the logger. I know it's only stored on changes, but given the potential flakiness of LIFX bulbs, the RSSI will change every polling interval for sure.

Djelibeybi committed:
Add a reboot button entity to the LIFX integration

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Move calling the device reboot method to the button entity

I had it under the coordinator for reasons that are no longer
relevent, but I didn't move the method itself.

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Added tests for the restart and identify buttons and fixed the light tests

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Fixed the integration instead of breaking the test

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
This is now rebased on top of the Button entity codebase.

Djelibeybi committed:
Fix the tests and update LIFXManager to handle multiple coordinators

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Added tests to cover the the sensors but not all of the coordinator

Ironically, I'm having trouble making tests that fail. Go figure.

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Now with tests for the sensor entities and 100% code coverage

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Seems the GitHub Actions mypy is stricter than the pre-commit one

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
@bdraco when you get a chance, could you review the tests I added? I'm fairly certain I picked the least efficient way of testing things, so would appreciate an experienced eye.

On that note, I'll probably stop adding entities until this lot is merged and work on replacing `aiolifx` with Photons or something closely resembling it.

Djelibeybi committed:
Update the aiolifx dependency and remove aiolifx_connection

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Add required method docstring

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Clean up the button entities from review feedback

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Rename the coordinators and coordination model

Signed-off-by: Avi Miller <me@dje.li>

bdraco reviewed (dismissed):
Please resolve conflicts.  Thanks 👍

Djelibeybi committed:
Fix up the types of the update coordinators

Signed-off-by: Avi Miller <me@dje.li>

bdraco commented:
I won't have time to do a proper review until tomorrow, but in the mean time I'll give feedback on the things I know will require refactoring

bdraco commented on tests/components/lifx/test_sensor.py:
```diff
@@ -0,0 +1,196 @@
+        await hass.config_entries.async_reload(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    lifx_coordination: LIFXCoordination = hass.data[DOMAIN][config_entry.entry_id]
```
We don't want to access `hass.data` in tests since thats considered  integration details.

patch the library instead and call async_fire_time_changed

https://developers.home-assistant.io/docs/development_testing?_highlight=writing&_highlight=tests#writing-tests-for-integrations

Djelibeybi replied:
I'm not sure I'm following. Can you point me towards an existing example of this, perhaps?

bdraco commented on tests/components/lifx/test_sensor.py:
```diff
@@ -0,0 +1,196 @@
+    )
+    assert states[1].attributes.get(ATTR_UNIT_OF_MEASUREMENT) == TIME_SECONDS
+
+    disabled_entities = [
+        entity_registry.async_update_entity(
+            entity_id, disabled_by=er.RegistryEntryDisabler.INTEGRATION
+        )
+        for entity_id in entity_ids
+    ]
```
This could be a helper

Djelibeybi replied:
As above, any chance you can point me towards an example?

bdraco replied:
Traveling this week but should have some time later in the week after jet lag recovery

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,254 @@
+    async def async_added_to_hass(self) -> None:
+        """Enable fetch of HEV data."""
+        setattr(self.coordinator, f"update_{self.entity_description.key}", True)
+        await self.coordinator.async_request_refresh()
```
```suggestion
```

`update_before_add` will already call `async_update`

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,254 @@
+    @callback
+    async def async_added_to_hass(self) -> None:
+        """Enable fetch of HEV data."""
+        setattr(self.coordinator, f"update_{self.entity_description.key}", True)
```
Please avoid `setattr`. Create a function on the coordinator instead

bdraco replied:
Better would be to have the function return a callable with CALLBACK_TYPE and then wrap the result in `self.async_on_remove` so you can drop `async_will_remove_from_hass`

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,254 @@
+        """Initialise a LIFX HEV sensor."""
+        super().__init__(coordinator, description)
+        self._attr_name = description.name
+        self.coordinator.update_method = cast(
+            Callable, f"self.coordinator._async_fetch_{description.key}"
+        )
```
Please avoid the cast here.

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,254 @@
+        """Update the uptime sensor attributes."""
+        self._attr_native_value = self.coordinator.uptime
+
+    @callback
+    async def async_added_to_hass(self) -> None:
+        """Enable updates for uptime sensor."""
+        self.coordinator.update_uptime = True
+        await self.coordinator.async_request_refresh()
+        await super().async_added_to_hass()
+
+    @callback
+    async def async_will_remove_from_hass(self) -> None:
+        """Disable updates for uptime sensor."""
+        self.coordinator.update_uptime = False
```
see feedback below

bdraco commented on homeassistant/components/lifx/coordinator.py:
```diff
@@ -183,3 +211,136 @@ async def async_set_color_zones(
+                if uptime is not None:
+                    self.uptime = uptime
+
+            except asyncio.TimeoutError as ex:
+                raise UpdateFailed(
+                    f"Failed to fetch uptime: {self.device.label} ({self.device.ip_addr})"
+                ) from ex
```
This could be a decorator instead since you are doing the same thing multiple places

Djelibeybi commented:
I'm on leave until next week, so I'll either have plenty of time for this or none at all. 😄  Worst case, I'll get back into it next Monday/Tuesday, as I'm away this weekend for sure.

Djelibeybi commented:
@bdraco I have some personal stuff I need to focus on this week, so I won't have much time to look at this until next week. It can certainly wait, but if you felt like tinkering on it, be my guest.

Djelibeybi committed:
Simplify the LIFX sensor entities after initial code review

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Refactored the HEV data into a binary sensor with extra attributes

This greatly simplifies the user interface and experience while still
providing sufficient sensor data to enable automation and frontend
dashboard content.

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
@bdraco I refactored the HEV sensors into a single binary sensor with extra state attributes which provides a much better user experience (IMO). At a glance, you can see if a HEV cycle is running (and firing automations on start/finish is possible) while still containing enough extra data for dashboard output or statistical analysis.

I realise this PR now has two platforms instead of one, but it would be super peachy keen if I don't have to split it into two seperate PRs.

Updated/new tests still to come.

Djelibeybi committed:
Fix existing tests to work with updated debug logging

Signed-off-by: Avi Miller <me@dje.li>

bdraco commented:
> @bdraco I refactored the HEV sensors into a single binary sensor with extra state attributes which provides a much better user experience (IMO). At a glance, you can see if a HEV cycle is running (and firing automations on start/finish is possible) while still containing enough extra data for dashboard output or statistical analysis.

Oh no, in general we are moving away from extra state attributes.  They can't be assigned state classes, can't use long term stats, can't be deduplicated in the database, are hard to discover, are more difficult to use in automations for first time users, etc.

> 
> I realise this PR now has two platforms instead of one, but it would be super peachy keen if I don't have to split it into two seperate PRs.
> 
> Updated/new tests still to come.

It would be better to split it into two PRs, but I think you already know that, or you wouldn't have asked.  If everything were super strait-forward here, it would probably be fine, but I don't want either of us to end up frustrated in the review process, so it is probably better to split it up.

Djelibeybi commented:
> Oh no, in general we are moving away from extra state attributes. They can't be assigned state classes, can't use long term stats, can't be deduplicated in the database, are hard to discover, are more difficult to use in automations for first time users, etc.

These are all the reasons why I switched to them for this, to be honest. But I guess I'll split them back out again and split this PR accordingly. Or close this one and open two new ones or something.

Djelibeybi commented:
Closing this PR and opening new PRs per platform. First new PR: https://github.com/home-assistant/core/pull/77535
